src/adaptation/icap/OptXact.cc

```diff
@@ -105,17 +106,35 @@ bool Adaptation::Icap::OptXact::parseResponse()
     debugs(93, 5, "have " << readBuf.length() << " bytes to parse" << status());
     debugs(93, DBG_DATA, "\n" << readBuf);
 
+    if (!replyParser)
+        replyParser = new ResponseParser();
+
+    Must(replyParser->parse(readBuf)); // throws on parse failure OR incomplete data
+    readBuf = replyParser->remaining(); // leave any unparsed bytes in the I/O buffer
+
+    /* We know the whole response is in parser now */
+    debugs(93, 2, "ICAP Server " << connection);
+    debugs(93, 2, "ICAP Server RESPONSE:\n---------\n" <<
+           replyParser->messageProtocol() << " " << replyParser->messageStatus() << " " << replyParser->reasonPhrase() << "\n" <<
+           replyParser->mimeHeader() <<
+           "----------");
+
     HttpReply::Pointer r(new HttpReply);
     r->protoPrefix = "ICAP/"; // TODO: make an IcapReply class?
+    r->sline.set(replyParser->messageProtocol(), replyParser->messageStatus());
 
-    if (!parseHttpMsg(r.getRaw())) // throws on errors
-        return false;
+    // parse headers
+    if (!r->parseHeader(*replyParser)) {
+        r->sline.set(replyParser->messageProtocol(), Http::scInvalidHeader);
+        debugs(93, 2, "error parsing response headers mime block");
+    }
```

Comment:
AFAICT, this replacement of the parseHttpMsg() call is incorrect; it does not follow parseHttpMsg() API/invariants. I do not know whether we should keep parseHttpMsg() or not, but I do not want to write parsing code using change requests. Have you tried preserving parseHttpMsg() calls?

Reply:
AFAICT the parseHttpMsg() attempted to parse the buffer and chopped out the bytes which appeared to be a valid response (status-line + headers block). That is all performed by the new liens 112-113 above
```C++
    Must(replyParser->parse(readBuf)); // throws on parse failure OR incomplete data
    readBuf = replyParser->remaining(); // leave any unparsed bytes in the I/O buffer
```

This if-statement is only doing the post-processing of the mime headers already identified.